### ISSUE TITLE
Try custom loss function for multilabel task

### ISSUE THREAD

**Issue by toddnief** [uchicago-dsi/good-food-purchasing#140]

**Comment by toddnief** [comment]
import torch
import torch.nn as nn

class CustomBCELoss(nn.Module):
    def __init__(self, pos_weight, fn_penalty):
        super(CustomBCELoss, self).__init__()
        self.pos_weight = pos_weight
        self.fn_penalty = fn_penalty
        self.bce_loss = nn.BCELoss(reduction='none')

    def forward(self, outputs, targets):
        loss = self.bce_loss(outputs, targets)
        
        # Apply class weights
        loss = loss * (targets * self.pos_weight + (1 - targets))
        
        # Apply higher penalties for false negatives
        fn_mask = (targets == 1) & (outputs < 0.5)
        loss = loss + fn_mask * self.fn_penalty

        return loss.mean()

# Example usage
pos_weight = torch.tensor([2.0])  # Adjust according to class distribution
fn_penalty = torch.tensor([1.5])  # Adjust according to specific needs

criterion = CustomBCELoss(pos_weight=pos_weight, fn_penalty=fn_penalty)

# Example tensors
outputs = torch.sigmoid(torch.randn(10))  # Predicted probabilities
targets = torch.randint(0, 2, (10,)).float()  # Ground truth labels

loss = criterion(outputs, targets)
print(loss)
